File: ruby/05_from_push_to_pull/lib/values_of_month.rb
class ValuesOfMonth
  attr_reader :balance

  def set_balance_and_average(balance, average_balance)
    @balance = balance
    @average_balance = average_balance
  end

  def average_balance
    Integer(@average_balance)
  end

  def calculate_values(date_of_month, transactions_of_month, preceding_balance)
    balance = preceding_balance
    ultimo = date_of_month.day

    average_balance = 0.0
    day_of_latest_balance = 1
    transactions_of_month.each do |transaction|
      day = transaction.date.day
      average_balance += calculate_proportional_balance(day_of_latest_balance, balance, day, ultimo)
      balance += transaction.amount
      day_of_latest_balance = day
    end
    average_balance += calculate_proportional_balance(day_of_latest_balance, balance, ultimo + 1, ultimo)
    set_balance_and_average(balance, average_balance)
  end

  private
  def calculate_proportional_balance(day_of_latest_balance, balance, day, days_in_month)
    counting_days = day - day_of_latest_balance
    return 0 if counting_days == 0
    rate = counting_days.to_f / days_in_month.to_f
    balance * rate
  end
end

File: csharp/06_FromPushToPull/ValuesOfMonth.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Common;

namespace FromPushToPull
{
    class ValuesOfMonth
    {
        public int Balance { get; set; }

        public int AverageBalance { get; set; }

        public void CalculateValues(DateTime dateOfMonth, IList<Transaction> transactionsOfMonth, int precedingBalance)
        {
            int balance = precedingBalance;
            int ultimo = dateOfMonth.Day;

            double averageBalance = 0;
            int dayOfLatestBalance = 1;
            foreach (Transaction transaction in transactionsOfMonth)
            {
                int day = transaction.Date.Day;
                averageBalance += CalculateProportionalBalance(dayOfLatestBalance, balance, day, ultimo);
                balance += transaction.Amount;
                dayOfLatestBalance = day;
            }

            averageBalance += CalculateProportionalBalance(dayOfLatestBalance, balance, ultimo + 1, ultimo);

            Balance = balance;
            AverageBalance = (int)averageBalance;
        }

        private double CalculateProportionalBalance(int dayOfLatestBalance, int balance, int day, int daysInMonth)
        {
            int countingDays = day - dayOfLatestBalance;
            if (countingDays == 0)
            {
                return 0;
            }
            double rate = (double)countingDays / daysInMonth;
            return (balance * rate);
        }
    }
}
